mbtiles/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+#![allow(clippy::missing_errors_doc)]
+
+extern crate core;
+
+use std::ffi::OsStr;
+use std::fmt::Display;
+use std::path::Path;
+use std::str::FromStr;
+
+use futures::TryStreamExt;
+use log::{debug, warn};
+use martin_tile_utils::DataFormat;
+use serde_json::{Value as JSONValue, Value};
+use sqlx::pool::PoolConnection;
+use sqlx::sqlite::SqlitePool;
+use sqlx::{query, Pool, Sqlite};
+use tilejson::{tilejson, Bounds, Center, TileJSON};
+
+#[derive(thiserror::Error, Debug)]
+pub enum MbtError {
+    #[error("SQL Error {0}")]
+    SqlError(#[from] sqlx::Error),
+
+    #[error(r"Inconsistent tile formats detected: {0:?} vs {1:?}")]
+    InconsistentMetadata(DataFormat, DataFormat),
+
+    #[error("No tiles found")]
+    NoTilesFound,
+}
+
+type MbtResult<T> = Result<T, MbtError>;
+
+#[derive(Clone, Debug)]
+pub struct Mbtiles {
+    filename: String,
+    pool: Pool<Sqlite>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Metadata {
+    pub tilejson: TileJSON,
+    pub id: String,
+    pub tile_format: DataFormat,
+    pub grid_format: Option<DataFormat>,
+    pub layer_type: Option<String>,
+    pub json: Option<JSONValue>,
+}
+
+impl Mbtiles {
+    pub async fn new(file: &Path) -> MbtResult<Self> {
+        // TODO: introduce a new error type for invalid file, instead of using lossy
+        let pool = SqlitePool::connect(&file.to_string_lossy()).await?;
+        let filename = file
+            .file_stem()
+            .unwrap_or_else(|| OsStr::new("unknown"))
+            .to_string_lossy()
+            .to_string();
+        Ok(Self { filename, pool })
+    }
+
+    fn to_val<V, E: Display>(&self, val: Result<V, E>, title: &str) -> Option<V> {
+        match val {
+            Ok(v) => Some(v),
+            Err(err) => {
+                let name = &self.filename;
+                warn!("Unable to parse metadata {title} value in {name}: {err}");
+                None
+            }
+        }
+    }
+
+    pub async fn get_metadata(&self) -> MbtResult<Metadata> {
+        let mut res = Metadata {
+            tilejson: tilejson! {
+                tiles: vec![String::new()],
```

Comment:
I think this adds an empty string in the tiles array.

Reply:
Thanks, I'll need to fix this, and also handle the x-forwarded-for header (there is a ticket for that). I'll have time tonight to look at it, in a few hrs

Reply:
That header works for me already. It places the correct host and protocol

Reply:
which header? We currently handle `x-rewrite-url`, which is a really bad one from the security perspective. See #504